gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/GraphManager.java

```diff
@@ -21,139 +21,98 @@
-        return traversalSources;
-    }
+    public Map<String, TraversalSource> getTraversalSources();
+
+    /**
+     * Get {@link TraversalSource} instance whose name matches {@link tsName}
+     *
+     * @return {@link TraversalSource} if exists, else null
+     */
 
+    public TraversalSource getTraversalSource(final String tsName);
     /**
      * Get the {@link Graph} and {@link TraversalSource} list as a set of bindings.
      */
-    public Bindings getAsBindings() {
-        final Bindings bindings = new SimpleBindings();
-        graphs.forEach(bindings::put);
-        traversalSources.forEach(bindings::put);
-        return bindings;
-    }
+
+    /**
+     * Add {@link TraversalSource} ts with name {@link String} tsName to
+     * {@link Map<String, TraversalSource>} returned by call to getTraversalSources()
+     */
+    public void addTraversalSource(final String tsName, final TraversalSource ts);
+
+    public Bindings getAsBindings();
 
     /**
      * Rollback transactions across all {@link Graph} objects.
      */
-    public void rollbackAll() {
-        graphs.entrySet().forEach(e -> {
-            final Graph graph = e.getValue();
-            if (graph.features().graph().supportsTransactions() && graph.tx().isOpen())
-                graph.tx().rollback();
-        });
-    }
+    public void rollbackAll();
 
     /**
      * Selectively rollback transactions on the specified graphs or the graphs of traversal sources.
      */
-    public void rollback(final Set<String> graphSourceNamesToCloseTxOn) {
-        closeTx(graphSourceNamesToCloseTxOn, Transaction.Status.ROLLBACK);
-    }
+    public void rollback(final Set<String> graphSourceNamesToCloseTxOn);
 
     /**
      * Commit transactions across all {@link Graph} objects.
      */
-    public void commitAll() {
-        graphs.entrySet().forEach(e -> {
-            final Graph graph = e.getValue();
-            if (graph.features().graph().supportsTransactions() && graph.tx().isOpen())
-                graph.tx().commit();
-        });
-    }
+    public void commitAll();
 
     /**
      * Selectively commit transactions on the specified graphs or the graphs of traversal sources.
      */
-    public void commit(final Set<String> graphSourceNamesToCloseTxOn) {
-        closeTx(graphSourceNamesToCloseTxOn, Transaction.Status.COMMIT);
-    }
+    public void commit(final Set<String> graphSourceNamesToCloseTxOn);
+
+    /**
+     * Implementation that allows for custom graph-opening implementations; if the {@link Map}
+     * tracking graph references has a {@link Graph} object corresponding to the {@link String} graphName,
+     * then we return that {@link Graph}-- otherwise, we use the custom {@link Supplier} to instantiate a
+     * a new {@link Graph}, add it to the {@link Map} tracking graph references, and return said {@link Graph}.
+     */
+    public Graph openGraph(final String graphName, final Supplier<Graph> supplier);
```

Comment:
I'm feeling like the second argument here could offer a little more flexibility if it were `BiSupplier<String,Graph>` where the first argument was the `graphName` passed back in on itself (or ultimately whatever name the `GraphManager` implementation used to key the created `Graph` instance. That could be a useful configuration parameter to something like Neo4j where it could use the graph name to uniquely create a data directory for example.